File: Strand/Collect/Backfiller.swift
import Foundation
import WhoopProtocol
import WhoopStore

// MARK: - BackfillStoreWriting protocol

/// The async subset the Backfiller needs. Plain async protocol (not @MainActor) so both the
/// real WhoopStore actor and a @MainActor SpyBackfillStore in tests can satisfy it.
protocol BackfillStoreWriting: AnyObject {
    @discardableResult
    func insert(_ streams: Streams, deviceId: String) async throws
        -> (hr: Int, rr: Int, events: Int, battery: Int,
            spo2: Int, skinTemp: Int, resp: Int, gravity: Int)
    func enqueueRawBatch(_ meta: RawBatchMeta, frames: [[UInt8]]) async throws
    func setCursor(_ name: String, _ value: Int) async throws
    func cursor(_ name: String) async throws -> Int?
}

extension WhoopStore: BackfillStoreWriting {}

// MARK: - Backfiller

/// Historical-offload state machine (idle / backfilling).
///
/// Per-chunk local safe-trim invariant:
///   decode known → await insert (decoded durable) →
///   await enqueueRawBatch (raw durable) →
///   await setCursor(strap_trim) →
///   ackTrim (link-layer confirmed ack to strap)
///
/// A chunk is forgotten only after decoded AND raw are both locally durable AND the ack
/// (.withResponse) is link-layer confirmed. Never waits on the server.
@MainActor
final class Backfiller {
    typealias Extractor = ([ParsedFrame], Int, Int) -> Streams

    private let store: BackfillStoreWriting
    private let deviceId: String
    /// Confirms one HISTORY_END chunk to the strap. Carries both the trim cursor (= first u32
    /// of end_data, used for the `strap_trim` cursor) and the 8-byte `end_data` (= the raw
    /// HISTORY_END metadata.data[10:18]) that the high-freq-sync ack form requires verbatim.
    private let ackTrim: (_ trim: UInt32, _ endData: [UInt8]) -> Void
    private let extract: Extractor
    /// Research toggle. When false (DEFAULT) no raw frames are persisted — the chunk's
    /// decoded streams are still durable and the trim is still acked (decoded is the product of
    /// record). Injected for tests; backed by UserDefaults in the production init site.
    private let enableRawCapture: Bool

    /// The clock reference set by BLEManager when GET_CLOCK confirms (required for decoding).
    var clockRef: ClockRef?

    /// True while a historical offload session is active.
    private(set) var isBackfilling = false

    /// Buffered data frames for the current open chunk (between START and END).
    private var chunk: [[UInt8]] = []
    /// Whether a START has been received and we're accumulating a chunk.
    private var chunkOpen = false
    /// Strap family for the current offload, set at begin(). Drives family-aware frame parsing (WHOOP 5/MG
    /// records sit at +4 offsets vs WHOOP 4.0) and the end_data slice the ack needs. Captured at begin()
    /// rather than init so it's correct even if the Backfiller was constructed before the strap was known.
    private var family: DeviceFamily = .whoop4

    /// Diagnostic sink (strap log). Surfaces historical records whose firmware layout we can't decode.
    private let log: ((String) -> Void)?
    /// Versions already reported this session, so the diagnostic logs each once (no spam).
    private var loggedUnmappedVersions: Set<Int> = []

    init(store: BackfillStoreWriting,
         deviceId: String,
         ackTrim: @escaping (_ trim: UInt32, _ endData: [UInt8]) -> Void,
         enableRawCapture: Bool = false,
         log: ((String) -> Void)? = nil,
         extract: @escaping Extractor = { extractHistoricalStreams($0, deviceClockRef: $1, wallClockRef: $2) }) {
        self.store = store
        self.deviceId = deviceId
        self.ackTrim = ackTrim
        self.enableRawCapture = enableRawCapture
        self.log = log
        self.extract = extract
    }

    /// Called by BLEManager when the strap signals a historical offload is beginning.
    /// chunkOpen starts TRUE: the high-freq-sync biometric replay streams records immediately and
    /// sends one HISTORY_START then repeated HISTORY_ENDs, so we must accumulate from the outset.
    func begin(family: DeviceFamily) {
        self.family = family
        isBackfilling = true
        chunk.removeAll(keepingCapacity: true)
        chunkOpen = true
    }

    /// Feed one raw BLE frame into the state machine. May trigger async store operations.
    func ingest(_ frame: [UInt8]) async {
        switch classifyHistoricalMeta(parseFrame(frame, family: family)) {
        case .start:
            isBackfilling = true
            chunk.removeAll(keepingCapacity: true)
            chunkOpen = true
        case .end(let unix, let trim):
            await finishChunk(unix: unix, trim: trim, endFrame: frame)
        case .complete:
            isBackfilling = false
            chunk.removeAll(keepingCapacity: true)
            chunkOpen = false
        case .other:
            if chunkOpen { chunk.append(frame) }
        }
    }

    /// The 8-byte `end_data` the high-freq-sync ack requires: metadata.data[10:18].
    /// metadata.data begins at frame[7] (after [type,seq,cmd]), so end_data = frame[17:25].
    /// trim cursor = the first u32 of end_data (data[10:14]). Returns nil if the frame is too
    /// short to contain the field (shouldn't happen for a real HISTORY_END, which is >=14 data
    /// bytes, but guards against a malformed frame).
    static func endData(from frame: [UInt8], family: DeviceFamily) -> [UInt8]? {
        // metadata.data begins at frame[7] (WHOOP4) / frame[11] (WHOOP5, the +4 puffin envelope); the
        // ack's end_data = data[10:18] → frame[17:25] (WHOOP4) or frame[21:29] (WHOOP5). The WHOOP5 slice
        // is verified on a real HISTORY_END (trim=112193 = frame[21..25]) in Whoop5HistoricalTests.
        let start = family == .whoop5 ? 21 : 17
        guard frame.count >= start + 8 else { return nil }
        return Array(frame[start..<(start + 8)])
    }

    /// Commit one HISTORY_END chunk: (persist decoded → enqueueRaw when present) → setCursor → ackTrim.
    /// Early-returns on any throw to preserve the safe-trim invariant.
    ///
    /// CRITICAL: high-freq-sync sends ONE HISTORY_START then REPEATED HISTORY_ENDs (a chunk-close
    /// every ~50 records). So we must ack EVERY end and keep accumulating afterwards — NOT close
    /// the chunk after the first. We snapshot+clear the accumulated frames but leave `chunkOpen`
    /// TRUE so the records following this END become the next chunk. An END with no accumulated
    /// records is still acked (it advances the strap's trim) — that's how the offload progresses.
    /// `endFrame` carries the 8-byte `end_data` the ack requires.
    private func finishChunk(unix: UInt32, trim: UInt32, endFrame: [UInt8]) async {
        guard let endData = Backfiller.endData(from: endFrame, family: family) else { return }

        let frames = chunk
        chunk.removeAll(keepingCapacity: true)   // next records accumulate into the next chunk

        if !frames.isEmpty {
            // type-47 HISTORICAL_DATA carries its OWN real-unix timestamp — extractHistoricalStreams
            // ignores the clock offset for it — so the historical offload does NOT need GET_CLOCK.
            // If the (device,wall) correlation isn't established yet (e.g. GET_CLOCK silent), fall back
            // to an identity ref (device==wall==now): the offset math becomes a no-op, type-47 still
            // decodes to correct wall time, and we can persist + ack + upload. The correlation is only
            // truly required to map REALTIME (type-40/43) device-epoch timestamps, never in a hist chunk.
            let ref = clockRef ?? { let now = Int(Date().timeIntervalSince1970); return ClockRef(device: now, wall: now) }()
            let parsed = frames.map { parseFrame($0, family: family) }
            // Diagnostic (#30): a historical record whose firmware version we don't have a field map for
            // bails out of decode entirely — no HR, no R-R, no GRAVITY — so sleep (which is gravity/
            // motion-driven) can never be computed from it, even though the offload "completes". Surface
            // each unmapped version once so the user's strap log reveals what their firmware emits.
            for p in parsed {
                guard let v = p.parsed["hist_version"]?.intValue,
                      p.parsed["heart_rate"] == nil,            // decoded nothing → unmapped layout
                      !loggedUnmappedVersions.contains(v) else { continue }
                loggedUnmappedVersions.insert(v)
                log?("Historical records use firmware layout v\(v), which NOOP doesn't decode yet — no motion data, so sleep can't be computed from the strap. Please report this (issue #30).")
            }
            let decoded = extract(parsed, ref.device, ref.wall)
            // Diagnostic (#77): the AGGREGATE silent-loss case — frames arrived but produced no rows at
            // all (CRC fail / unmapped layout / out-of-range timestamp), so this chunk persists nothing
            // yet still acks below and the strap trims past it. The per-version log above only catches
            // unmapped layouts; this catches CRC drops too. Observability only — behaviour unchanged
            // (not acking would wedge the offload on a re-send loop). Surfaces in the user's strap log.
            if decoded.isEmpty {
                log?("Backfill: \(frames.count) frame(s) decoded to 0 rows (trim=\(trim)) — dropped (CRC/layout/timestamp); nothing persisted for this chunk.")
                // #91: dump a hex sample of the rejected frames so an unmapped firmware's record
                // layout can be mapped from a user's strap log — the count alone can't be decoded.
                for (i, f) in frames.prefix(3).enumerated() {
                    let hex = f.prefix(64).map { String(format: "%02x", $0) }.joined()
                    log?("Backfill: rejected frame[\(i)] \(f.count)B: \(hex)\(f.count > 64 ? "…" : "")")
                }
            }
            do { try await store.insert(decoded, deviceId: deviceId) } catch { return }

            // RAW: only persisted when the research toggle is ON. Default OFF → decoded-only; the
            // chunk is still durably committed (decoded) so the trim is safe to advance + ack.
            if enableRawCapture {
                let meta = RawBatchMeta(
                    batchId: "hist-\(deviceId)-\(trim)",
                    deviceId: deviceId,
                    clockRef: ref,
                    capturedAt: Int(Date().timeIntervalSince1970),
                    startTs: ref.wall,
                    endTs: ref.wall,
                    frameCount: frames.count,
                    byteSize: frames.reduce(0) { $0 + $1.count })
                do { try await store.enqueueRawBatch(meta, frames: frames) } catch { return }
            }
        }
        do { try await store.setCursor("strap_trim", Int(trim)) } catch { return }

        ackTrim(trim, endData)
    }

    /// Called when a backfill watchdog timer fires (strap went silent mid-offload).
    /// Clears state without acking — the chunk was never durably committed.
    func timeoutFired() {
        isBackfilling = false
        chunk.removeAll(keepingCapacity: true)
        chunkOpen = false
    }
}

File: android/app/src/main/java/com/noop/ble/Backfiller.kt
package com.noop.ble

import android.content.Context
import com.noop.data.StreamBatch
import com.noop.data.WhoopRepository
import com.noop.protocol.DeviceFamily
import com.noop.protocol.Framing
import com.noop.protocol.HistoricalMeta
import com.noop.protocol.classifyHistoricalMeta
import com.noop.protocol.extractHistoricalStreams
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

/**
 * Historical-offload state machine (idle / backfilling).
 *
 * Direct port of the macOS Swift `Backfiller` (Strand/Collect/Backfiller.swift). It consumes the
 * METADATA frames of an offload — HISTORY_START / repeated HISTORY_END / HISTORY_COMPLETE —
 * accumulating the type-47 records between them into chunks and committing each chunk durably.
 *
 * Per-chunk local safe-trim invariant (unchanged from Swift):
 *   decode known -> persist decoded (durable) -> persist the strap_trim cursor -> ack the trim to
 *   the strap (link-layer confirmed write).
 * A chunk is forgotten by the strap only after its decoded rows are locally durable AND the trim
 * cursor is persisted AND the ack write is confirmed. The phone NEVER waits on a server (there is
 * none — Strand is fully on-device).
 *
 * CRITICAL behaviour preserved from Swift: a high-freq-sync offload sends ONE HISTORY_START then
 * REPEATED HISTORY_ENDs (a chunk-close every ~50 records). So we ack EVERY end and keep
 * accumulating afterwards — we snapshot+clear the accumulated frames on each END but leave the
 * chunk OPEN so subsequent records become the next chunk. An END with no accumulated records is
 * still acked (it advances the strap's trim) — that is how the offload progresses.
 *
 * CONCURRENCY: [ingest] is `suspend` and serialised by [mutex] so START/data/END chunk assembly is
 * never reordered, matching the Swift serial-drain task. The owning [WhoopBleClient] feeds frames
 * in arrival order from a single drain coroutine.
 *
 * RAW CAPTURE: the Swift Backfiller optionally persists raw frames (research toggle, default OFF).
 * The Android data layer has no raw-frame outbox table, so raw persistence is intentionally omitted
 * here — decoded rows are the product of record and are still durably committed before the trim is
 * advanced, exactly as in the Swift default (raw-off) configuration. See the FLAG in the port notes.
 */
class Backfiller(
    private val repository: WhoopRepository,
    private val deviceId: String,
    private val cursorStore: TrimCursorStore,
    /**
     * Confirms one HISTORY_END chunk to the strap. Carries both the trim cursor (first u32 of
     * end_data, persisted as the `strap_trim` cursor) and the verbatim 8-byte `end_data` (the raw
     * HISTORY_END metadata.data[10:18]) the high-freq-sync ack form requires.
     */
    private val ackTrim: (trim: Long, endData: ByteArray) -> Unit,
    /**
     * Fires after a chunk's decoded rows are durably committed AND acked — i.e. real new data just
     * landed. Lets the client schedule on-device scoring right away instead of leaving fresh history
     * invisible until the next 15-min analysis tick. Empty chunks (metadata-only ENDs) don't fire.
     * (#78 fork)
     */
    private val onChunkCommitted: (StreamBatch) -> Unit = {},
    /**
     * Diagnostic sink into the strap log. Lets [finishChunk] surface a chunk that arrived with frames
     * but decoded to ZERO rows — the otherwise-invisible silent-data-loss case (frames failing CRC /
     * an unmapped layout are dropped, the chunk looks empty, the trim acks past them). Without this a
     * "zero data" strap log shows healthy "acked chunk" lines while data is being discarded (#77). */
    private val log: (String) -> Unit = {},
    /**
     * The (device, wall) clock reference. type-47 records carry their OWN real unix timestamp so
     * the offset is a no-op for them; this is supplied only for the REALTIME_RAW_DATA fallback and
     * to mirror the Swift signature. Defaults to an identity ref (device == wall == now): the Swift
     * Backfiller falls back to exactly this when GET_CLOCK is silent, and type-47 still decodes to
     * correct wall time. Settable by [WhoopBleClient] if a real correlation lands.
     */
    var clockRef: ClockRef = ClockRef.identityNow(),
) {

    /**
     * Strap family for the CURRENT offload, set at [begin] — drives the family-aware frame parse
     * (5/MG inner record is +4) and the +4 end_data slice. The Backfiller is constructed once at
     * client init (before the family is known), so this is settable per-offload rather than a
     * constructor arg. Mirrors Swift `Backfiller.family` set in `begin(family:)`. (#78)
     */
    private var family: DeviceFamily = DeviceFamily.WHOOP4

    /** True while a historical offload session is active. */
    @Volatile
    var isBackfilling = false
        private set

    /** Serialises the suspend [ingest] calls so chunk boundaries are never crossed concurrently. */
    private val mutex = Mutex()

    /** Guards the [chunk]/[chunkOpen] mutations (the only cross-thread state: ingest vs begin/timeout). */
    private val chunkLock = Any()

    /** Buffered data frames for the current open chunk (between START and the next END). */
    private val chunk = ArrayList<ByteArray>()

    /** Whether a START has been received and we're accumulating a chunk. */
    private var chunkOpen = false

    /**
     * Called by [WhoopBleClient] when the strap signals a historical offload is beginning.
     * chunkOpen starts TRUE: the biometric replay streams records immediately and sends one
     * HISTORY_START then repeated HISTORY_ENDs, so we must accumulate from the outset.
     * Port of Swift `begin()`.
     */
    fun begin(family: DeviceFamily = DeviceFamily.WHOOP4) {
        this.family = family
        isBackfilling = true
        synchronized(chunkLock) {
            chunk.clear()
            chunkOpen = true
        }
    }

    /**
     * Feed one complete (reassembled) BLE frame into the state machine. Suspends while a chunk is
     * persisted so chunk boundaries are never crossed concurrently. Port of Swift `ingest(_:)`.
     */
    suspend fun ingest(frame: ByteArray) {
        mutex.withLock {
            when (val meta = classifyHistoricalMeta(Framing.parseFrame(frame, family))) {
                is HistoricalMeta.Start -> {
                    isBackfilling = true
                    synchronized(chunkLock) {
                        chunk.clear()
                        chunkOpen = true
                    }
                }
                is HistoricalMeta.End -> finishChunk(meta.unix, meta.trim, frame)
                is HistoricalMeta.Complete -> {
                    isBackfilling = false
                    synchronized(chunkLock) {
                        chunk.clear()
                        chunkOpen = false
                    }
                }
                is HistoricalMeta.Other -> synchronized(chunkLock) { if (chunkOpen) chunk.add(frame) }
            }
        }
    }

    /**
     * Commit one HISTORY_END chunk: persist decoded -> persist strap_trim cursor -> ack the trim.
     * Early-returns on any failure to preserve the safe-trim invariant (never ack data we failed to
     * store). Port of Swift `finishChunk(unix:trim:endFrame:)`.
     *
     * We snapshot+clear the accumulated frames but leave [chunkOpen] TRUE so the records following
     * this END become the next chunk. An END with no records is still acked (advances the trim).
     */
    private suspend fun finishChunk(unix: Long, trim: Long, endFrame: ByteArray) {
        val endData = endData(endFrame, family) ?: return

        val frames = synchronized(chunkLock) {
            val snapshot = ArrayList(chunk)
            chunk.clear() // next records accumulate into the next chunk
            snapshot
        }

        var committed: StreamBatch? = null
        if (frames.isNotEmpty()) {
            val ref = clockRef
            val decoded = extractHistoricalStreams(frames, ref.device, ref.wall, family)
            // DIAGNOSTIC (#77): frames arrived but produced no rows — they were dropped (CRC fail /
            // unmapped layout / out-of-range timestamp), so this chunk persists nothing yet still acks
            // below and the strap trims past it. Surface it loudly so a "zero data" strap log reveals
            // the silent loss instead of looking healthy. (Observability only — behaviour unchanged
            // pending a confirmed root cause; not acking here would wedge the offload on a re-send loop.)
            if (decoded.isEmpty) {
                log("Backfill: WARNING ${frames.size} frame(s) decoded to 0 rows (trim=$trim) — dropped (CRC/layout/timestamp); nothing persisted for this chunk")
                // #91: dump a hex sample of the rejected frames so an unmapped firmware's record
                // layout can be mapped from a user's strap log — the count alone can't be decoded.
                frames.take(3).forEachIndexed { i, f ->
                    val hex = f.take(64).joinToString("") { "%02x".format(it) }
                    log("Backfill: rejected frame[$i] ${f.size}B: $hex${if (f.size > 64) "…" else ""}")
                }
            }
            try {
                repository.insert(decoded, deviceId) // DECODED FIRST (durable)
                committed = decoded
            } catch (t: Throwable) {
                return // do NOT advance/ack — chunk was never durably committed
            }
        }

        // Persist the trim cursor BEFORE acking (so a crash between persist and ack still resumes
        // from the right place). Stored via [TrimCursorStore] because the Room schema has no cursor
        // table — see the port FLAG. trim is a u32 carried as Long (unsigned-safe).
        try {
            cursorStore.set(STRAP_TRIM_CURSOR, trim)
        } catch (t: Throwable) {
            return
        }

        ackTrim(trim, endData)
        committed?.takeIf { !it.isEmpty }?.let(onChunkCommitted)
    }

    /**
     * Called when a backfill watchdog timer fires (strap went silent mid-offload). Clears state
     * WITHOUT acking — the open chunk was never durably committed. Port of Swift `timeoutFired()`.
     */
    fun timeoutFired() {
        isBackfilling = false
        synchronized(chunkLock) {
            chunk.clear()
            chunkOpen = false
        }
    }

    companion object {
        /** Cursor name for the strap's safe-trim watermark. Matches the Swift `setCursor("strap_trim", ...)`. */
        const val STRAP_TRIM_CURSOR = "strap_trim"

        /**
         * The 8-byte `end_data` the high-freq-sync ack requires: metadata.data[10:18]. The inner
         * record begins at frame[7] on WHOOP4 (end_data = frame[17:25]) and at frame[11] on WHOOP5/MG
         * (the +4 puffin envelope → end_data = frame[21:29]). The trim cursor is the first u32 of
         * end_data. Returns null if the frame is too short. Verified against a real WHOOP5 HISTORY_END
         * (trim=112193 at frame[21:25]); port of Swift `Backfiller.endData(from:family:)`. (#78)
         */
        fun endData(frame: ByteArray, family: DeviceFamily): ByteArray? {
            val start = if (family == DeviceFamily.WHOOP5) 21 else 17
            if (frame.size < start + 8) return null
            return frame.copyOfRange(start, start + 8)
        }
    }
}

/**
 * A (device-epoch, wall-clock) correlation in unix seconds. Android analog of the Swift `ClockRef`.
 * type-47 historical records carry real unix timestamps, so the identity ref (device == wall) makes
 * the offset math a no-op while still decoding correct wall time — the same fallback the Swift
 * Backfiller uses when GET_CLOCK is silent.
 */
data class ClockRef(val device: Int, val wall: Int) {
    companion object {
        fun identityNow(): ClockRef {
            val now = (System.currentTimeMillis() / 1000L).toInt()
            return ClockRef(device = now, wall = now)
        }
    }
}

/**
 * Durable key/value cursor store. The macOS Backfiller persists `strap_trim` via the GRDB store's
 * cursor table; the Android Room schema has no cursor table (see Entities.kt — no cursor entity),
 * so this small SharedPreferences-backed store provides the equivalent durability WITHOUT touching
 * the Room schema or the build/manifest.
 *
 * FLAG (uncertain / divergence from macOS): on the Swift side the cursor lives in the same SQLite
 * file as the decoded rows, so cursor and rows commit/back-up atomically together. Here the cursor
 * lives in SharedPreferences, separate from the Room DB. The safe-trim ORDERING is preserved
 * (decoded rows are inserted and durable before the cursor is written, and the cursor is written
 * before the ack), so the worst case is a redundant re-offload of an already-stored chunk after a
 * crash — never data loss — because the decoded inserts are idempotent by natural key. If a Room
 * `cursor` table is later added, swap this implementation for a DAO-backed one.
 */
interface TrimCursorStore {
    suspend fun set(name: String, value: Long)
    suspend fun get(name: String): Long?
}

/** Default [TrimCursorStore] backed by a private SharedPreferences file. */
class PrefsTrimCursorStore(context: Context) : TrimCursorStore {
    private val prefs = context.applicationContext
        .getSharedPreferences("noop_backfill_cursors", Context.MODE_PRIVATE)

    override suspend fun set(name: String, value: Long) {
        // commit() (synchronous) so durability is established before we ack the strap.
        prefs.edit().putLong(name, value).commit()
    }

    override suspend fun get(name: String): Long? =
        if (prefs.contains(name)) prefs.getLong(name, 0L) else null
}
